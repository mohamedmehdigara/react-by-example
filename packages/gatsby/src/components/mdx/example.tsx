import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
