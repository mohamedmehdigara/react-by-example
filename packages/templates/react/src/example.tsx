import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
