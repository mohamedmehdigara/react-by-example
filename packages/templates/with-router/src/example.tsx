import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
