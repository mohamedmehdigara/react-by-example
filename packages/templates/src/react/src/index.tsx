import React from 'react'
import ReactDOM from 'react-dom'
import example from './example'

ReactDOM.render(<>{example}</>, document.getElementById('root'))
