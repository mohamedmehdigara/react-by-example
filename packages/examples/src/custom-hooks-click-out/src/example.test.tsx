import { cleanup, render, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React, { FC, useCallback, useState } from 'react'
import { Dropdown, useClickOut } from './example'

describe('useClickOut', () => {
  let component: RenderResult
  let list: HTMLElement
  let img: HTMLElement
  let button: HTMLElement
  let heading: HTMLElement
  let link: HTMLElement

  const ClickOutCounter: FC = () => {
    const [count, setCount] = useState(0)
    const decrement = useCallback(() => setCount((x) => x - 1), [])
    const increment = useCallback(() => setCount((x) => x + 1), [])

    return (
      <>
        <ul />
        <img alt="" />
        <button type="button" onClick={increment} {...useClickOut(decrement)}>
          click here to increment
        </button>
        <h1>click anywhere else to decrement</h1>
        <a href="/">{count}</a>
      </>
    )
  }

  beforeEach(() => {
    component = render(<ClickOutCounter />)
    list = component.getByRole('list')
    img = component.getByRole('img')
    button = component.getByRole('button')
    heading = component.getByRole('heading')
    link = component.getByRole('link')
  })

  it('should show 0', () => {
    expect(link).toHaveTextContent('0')
  })

  describe('when clicking button', () => {
    beforeEach(() => {
      userEvent.click(button)
      userEvent.click(button)
      userEvent.click(button)
      userEvent.click(button)
      userEvent.click(button)
    })

    it('should show 5', () => {
      expect(link).toHaveTextContent('5')
    })

    describe('when clicking list', () => {
      beforeEach(() => {
        userEvent.click(list)
      })

      it('should show 4', () => {
        expect(link).toHaveTextContent('4')
      })

      describe('when clicking img', () => {
        beforeEach(() => {
          userEvent.click(img)
        })

        it('should show 3', () => {
          expect(link).toHaveTextContent('3')
        })

        describe('when clicking heading', () => {
          beforeEach(() => {
            userEvent.click(heading)
          })

          it('should show 2', () => {
            expect(link).toHaveTextContent('2')
          })

          describe('when clicking link', () => {
            beforeEach(() => {
              userEvent.click(link)
            })

            it('should show 1', () => {
              expect(link).toHaveTextContent('1')
            })
          })
        })
      })
    })
  })
})

describe('Dropdown', () => {
  let component: RenderResult
  let toggle: HTMLElement

  beforeEach(() => {
    component = render(<Dropdown title="test">content</Dropdown>)
    toggle = component.getByRole('button')
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  it('should not show content', () => {
    expect(component.container).not.toHaveTextContent(/content/)
  })

  describe('when opened', () => {
    beforeEach(() => {
      userEvent.click(toggle)
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })

    it('should show content', () => {
      expect(component.container).toHaveTextContent(/content/)
    })

    describe('when closed using toggle', () => {
      beforeEach(() => {
        userEvent.click(toggle)
      })

      it('should not show content', () => {
        expect(component.container).not.toHaveTextContent(/content/)
      })
    })

    describe('when closed by clicking outside', () => {
      beforeEach(() => {
        const aside = render(<aside />).getByRole('complementary')
        userEvent.click(aside)
      })

      it('should not show content', () => {
        expect(component.container).not.toHaveTextContent(/content/)
      })
    })
  })
})
