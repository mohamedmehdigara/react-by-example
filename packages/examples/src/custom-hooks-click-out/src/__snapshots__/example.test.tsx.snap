// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <button
      class="toggle"
      type="button"
    >
      test
       
      ðŸ”¼
    </button>
  </div>
</div>
`;

exports[`Dropdown when opened should snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <button
      class="toggle"
      type="button"
    >
      test
       
      ðŸ”½
    </button>
    <div
      class="content"
    >
      content
    </div>
  </div>
</div>
`;
